### PR TITLE
Expose RSS-feed URL to browser

### LINKED ISSUE #423: %F placeholder for feed's URL in `browser` setting
A user [over at out mailing list](https://groups.google.com/forum/#!topic/newsboat/fZ2_7iE85F8) has a macro for deleting feeds from urls file. For that, they need the url of the currently selected feed. They noticed that `%u` doesn't always work, which is expected since it is feed's "alternative" URL, not the one by which it is fetched.

To aid such use cases, I suggest adding a new placeholder, `%U`, that will be replaced by the URL as it is in the urls file.

### PR DESCRIPTION
**Pull request by foxdevilwild**
Introduces a new '%U' formatting option for the 'browser' config
variable, which exposes the RSS-feed's URL.

Closes #423

**Note**: The relevant parts of the documentation seem to be gitignored?

### CODE REVIEW

File: src/view.cpp [hunk 1]
```diff
@@ -390,28 +390,25 @@ void View::open_in_pager(const std::string& filename)
...
-		std::string newurl;
-		newurl = utils::replace_all(url, "'", "%27");
-		newurl.insert(0, "'");
-		newurl.append("'");
-		fmt.register_fmt('u', newurl);
+		fmt.register_fmt('u', utils::quote_url_for_cmdline(itemurl));
+		fmt.register_fmt('U', utils::quote_url_for_cmdline(feedurl));
 		cmdline = fmt.do_format(browser, 0);
 	} else {
 		if (browser != "")
```
[the remark this answers is not in the document]
Comment (foxdevilwild) [reply src/view.cpp]: Sure

File: src/feedlistformaction.cpp [hunk 2]
```diff
@@ -230,6 +226,10 @@ void FeedListFormAction::process_operation(Operation op,
 							"FeedListFormAction: cannot open feed in browser "
```
[the remark this answers is not in the document]
Comment (foxdevilwild) [reply src/feedlistformaction.cpp]: Done

**Comment by foxdevilwild** [comment]
Thanks for the feedback. I've implemented your suggestions to the C++ code. From a first glance I think I'll be able to convert quote_url_for_cmdline to rust, but I'm afraid I won't get around to it until the weekend.

I'm also having trouble coming up with an example macro usage. Personally, I use it to call into custom scripts, which are probably not a great fit to be added to the docs. The user who suggested the change wants to use it to remove lines from `newsboat/urls`, but I feel uneasy about adding potentially destructive commands to the documentation. What do you think?

**Comment by Minoru** [comment]
> I'm afraid I won't get around to it until the weekend

That's totally fine, there is no hurry. The next scheduled release is at the end of March, there are plenty of weekends until then :) (I plan a bugfix release soon-ish, but it is not blocked by this PR.)

Regarding the macro example: can't come up with a good one, either. Just a paragraph on the difference between `%u` and `%U`? If you don't understand it well enough, I can merge this as-is and write the docs myself, later.

File: doc/configcommands.dsv [hunk 3]
```diff
@@ -8,7 +8,7 @@ bind-key||<key> <operation> [<dialog>]||n/a||Bind key <key> to <operation>. This
...
 bookmark-cmd||<command>||""||If set, then <command> will be used as bookmarking plugin. See the documentation on bookmarking for further information.||bookmark-cmd "~/bin/delicious-bookmark.sh"
 bookmark-interactive||[yes/no]||no||If set to `yes`, then the configured bookmark command is an interactive program.||bookmark-interactive yes
-browser||<command>||%BROWSER, otherwise lynx||Set the browser command to use when opening an article in the browser. If BROWSER environment variable is set, it will be used as the default browser, otherwise lynx will be used. Any occurrences of `%u` in <command> will be replaced by a URL in single quotes.||browser "w3m %u"
+browser||<command>||%BROWSER, otherwise lynx||Set the browser command to use when opening an article in the browser. If BROWSER environment variable is set, it will be used as the default browser, otherwise lynx will be used. Any occurrences of `%u` and `%U` in <command> will be replaced by the article's URL or the feed's URL, respectively, enclosed in single quotes.||browser "w3m %u"
```
Comment (Minoru) [inline doc/configcommands.dsv]: This is a good start, but it doesn't cover the case when `%u` is used and user tries to open a feed. [In that case](https://github.com/newsboat/newsboat/pull/437/files#diff-e953778e3f493f71c3eaec051e841374R230), `%u` is either the "alternative URL" of the feed, or its real URL if the alternative is not known. I find it quite tricky to put into words succinctly; feel free to have a try at it :) Or just roll back this change and I'll try myself, later.

File: rust/libnewsboat/src/utils.rs [hunk 4]
```diff
@@ -595,6 +595,17 @@ mod tests {
         );
     }
 
+    #[test]
+    fn t_quote_url_for_cmdline() {
+        assert_eq!(quote_url_for_cmdline("".to_string()), "''");
+        assert_eq!(quote_url_for_cmdline("'".to_string()), "'%27'");
+        assert_eq!(quote_url_for_cmdline("'one'two'".to_string()), "'%27one%27two%27'");
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: This line is too long, as `cargo fmt` [points out](https://travis-ci.org/newsboat/newsboat/jobs/490908516#L475). Please run `cargo fmt` and commit the result.

Comment (foxdevilwild) [reply rust/libnewsboat/src/utils.rs]: Done

File: doc/configcommands.dsv (hunk 3 above)
Comment (foxdevilwild) [reply doc/configcommands.dsv]: I think we should add a paragraph (and/or table) explaining the details somewhere else and just provide a brief example here. I'll see what I can come up with.

Comment (Minoru) [reply doc/configcommands.dsv]: This table is a reference, there can be only one example (in the last column). A separate table idea isn't that bad, though I'm not sure where to put it. Just after this "commands" table perhaps?

Comment (foxdevilwild) [reply doc/configcommands.dsv]: Yeah, the example provided is fine, I should've been clearer. I was thinking about something along the lines of:

```
[...] Any occurences of '%u' and '%U' in <command> will be replaced according to [insert link to  the browser formatting options]. If <command> contains neither '%u' nor '%U' newsboat will behave as though ' %u' was appended to <command>, implicitly.
````

Then, put a table at the end of the docs along with the other formatting options tables.

Comment (Minoru) [reply doc/configcommands.dsv]: Sounds good, I'd merge that!

Comment (Minoru) [reply doc/configcommands.dsv]: Ping. If you finished these changes, can you push them? No rush, just want to make sure you haven't abandoned this yet.

**Comment by foxdevilwild** [comment]
Hey Minoru. Apologies for the delay, I haven't abandoned the PR. The docs are still unfinished, but I hope to get around to it tomorrow.

**Comment by foxdevilwild** [comment]
I've made the discussed changes to the docs, and it seems to be working for the most part. However, when building, there is a validation error checking `doc/manpage-newsboat.xml`, which I am unable to resolve.

The failing check is:
`xmllint --nonet --noout --valid "doc/manpage-newsboat.xml"` 

It fails with:
`element link: validity error : IDREF attribute linkend references an unknown ID "browser-format-table"`

Looking at the file confirms that `browser-format-table` is indeed not defined there, however it _is_ defined in `doc/newsboat.txt` and when looking at `doc/xhtml/newboat.html` the link works as intended.

Hopefully you can point me in the right direction?

**Comment by Minoru** [comment]
`doc/manpage-newsboat.txt` is not included in `doc/configcommands.dsv`, so Asciidoc can't resolve the link. I believe our current practice is simply not to use links in `doc/configcommands.dsv`. That's quite lame, and that's one of the reasons why I filed #441. When that issue is fixed, we will include the DSV in the manpage, and will be able to cross-reference whatever we want.

File: doc/configcommands.dsv [hunk 5]
```diff
@@ -8,7 +8,7 @@ bind-key||<key> <operation> [<dialog>]||n/a||Bind key <key> to <operation>. This
...
 bookmark-interactive||[yes/no]||no||If set to `yes`, then the configured bookmark command is an interactive program.||bookmark-interactive yes
-browser||<command>||%BROWSER, otherwise lynx||Set the browser command to use when opening an article in the browser. If BROWSER environment variable is set, it will be used as the default browser, otherwise lynx will be used. Any occurrences of `%u` and `%U` in <command> will be replaced by the article's URL or the feed's URL, respectively, enclosed in single quotes.||browser "w3m %u"
+browser||<command>||%BROWSER, otherwise lynx||Set the browser command to use when opening an article in the browser. If BROWSER environment variable is set, it will be used as the default browser, otherwise lynx will be used. Any occurrences of `%u` and `%U` in <command> will be replaced according to the <<browser-format-table,browser format identifiers>>. If neither `%u` nor `%U` are included, newsboat will behave as though " `%u`" had been implicitly appended to <command>.||browser "w3m %u"
```
Comment (Minoru) [inline doc/configcommands.dsv]: > newsboat will behave as though " `%u`" had been implicitly appended to <command>

This should use single quotes, not double quotes.

I also believe that both `%u` and `%U` are always enclosed in single quotes; that was mentioned in the previous version of the doc, but is missing here.

**Comment by foxdevilwild** [comment]
Hey Minoru, just letting you know that I haven't abandoned the Issue. My HDD died and I'm currently left with only a mobile device until the replacement arrives (hopefully tomorrow). Apologies again for leaving this unresolved for the time being. I'll promise to get the PR ready as soon as possible.

**Comment by Minoru** [comment]
Oh, dead HDD sucks! Hope you had a backup.

Don't rush it, and don't even worry about the PR—the next release is at the end of March, so you've got plenty of time. And thanks for checking in; I wasn't sure if I should ping you after a week or wait a bit more.

**Comment by Minoru** [comment]
Hey @foxdevilwild, how is it going? Wanna push this over the finish line, or should I take over?

**Comment by Minoru** [comment]
I still haven't gotten around to finishing this, so I'm un-assigning myself. This PR is now free for the taking; please push it over the finish line!

**Comment by Minoru** [comment]
I believe this is easier to re-do from scratch than to rebase onto current master. We changed the way we escape URLs on the command line, and also reformatted a lot of code.

@foxdevilwild, no hard feelings for not finishing this—life happens. Your PRs are still welcome ;)

**Pull request closed without merging**